lib/xapi-stdext-date/date.ml

```diff
@@ -38,35 +28,86 @@ let rfc822_of_float x =
 
 let rfc822_to_string x = x
 
+(* ==== ISO8601/RFC3339 ==== *)
+type old_time = string
+type error = string
+type p_time = Ptime.t * Ptime.tz_offset_s option
+
+(** the name doesn't make much sense anymore, but is kept for compatibility reasons
+  * if string is received in rfc3339 format, timezone will be preserved between to/from string
+    but not once converted to a float
+*)
+type iso8601 =
```

Comment:
I like the design but like these names better:
```
type iso8601 =
  | Time of p_time
  | Legacy of string * error
```

I am a bit confused why an error is part of this data type. Should this not be orthogonal?

Reply:
I think this is due to a mismatch between parsing and validation that happens in the previous design.

There was no parsing whatsoever, allowing any string to be converted, and then there was a validation further down the line using `assert_utc`.

I'd rather parse the string into a utc-based timedate (Ptime, float) and fix what this change of behaviour breaks rather than trying to maintain all the information to have `of_string . to_string = id`